# Homework 8

## FINM 36700 - 2024

### UChicago Financial Mathematics

* Mark Hendricks
* hendricks@uchicago.edu


## HBS Case

### Long-Term Capital Management

---

# Section 1: Conceptual Issues for LTCM

### 1.1 Describe LTCM’s investment strategy with regard to the following aspects:
- **Securities traded:**

<span style="color: lightblue;">
They traded fixed income bonds (with prefeence for those with little or no default risk - that is, mostly developed-market sovereign debt and high-grade corporate bonds), fixed-rate residential mortages backed or implicitly backed by the government (FNMA, FHLMC and GNMA). Put and call options, and stocks. The financing (or short position) was tipically done with repurchase agreements, interest-rate swap agreements, and total-return swaps.
</span>

- **Trading frequency**

<span style="color: lightblue;"> 
Since most strategies were convergence and relatie values, the trading frequency depended on the time their trades took to converge, or not (in which case, the trade would typically be held to maturity), so ranging from weeks to years.
</span>

- **Skewness** (Do they seek many small wins or a few big hits?)

<span style="color: lightblue;"> It had a few big hits and small or slightly negative frequent returns, so _positively skewness_</span>

- **Forecasting** (What is behind their selection of trades?)

<span style="color: lightblue;">Their trades were almost all win/win situations, in which if their base case happened they would make money. If it did not happened, they would still be able to increase their position or maintain the trade until maturity, where it would provide positive returns. Still, some of the trades had small negative returns in case the base case did not happen, but because their correlation with the rest of portfolio was almost null, it would add little incremental risk to the portfolio, and in case the base case ocurred, the return would be substantial in because of the fact that if the spread converges, they make money. If it diverges, the trade becomes even more attractive, as convergence is still expected at a future date.
</span>

### 1.2 What are LTCM’s biggest advantages over its competitors?

- <span style="color: lightblue;"> It was able to finance positions at a lower cost. Since its trades rquired minimal, if any, initial outlay of capital, it could leverage considerably more than its competitors.
- <span style="color: lightblue;"> It had a diversified portfolio, which meant that uncorrelated trades like risk-arbitrage added little incremental risk to the overall portfolio, so the company could engage into those with smaller expected returns than pure risk-arbitrage funds.
- <span style="color: lightblue;"> It had a very large size, which allowed it to receive good prices in block trades.
</span>


### 1.3 The case discusses four types of funding risk facing LTCM.
#### Briefly discuss specific ways in which LTCM manages each of these risks:
- **Collateral haircuts** 

<span style="color: lightblue;">
For future financing needs, where unkown haircuts might be applied, LTCM assumed that financing haircuts could increase from 0% to 2% on German government bonds, 0% to 5% on Italian government bonds, and that it might face haircuts as high as 10% on instruments such as AAA-rated commercial mortgages which the firm presently could finance with 2% haircuts.
</span>

- **Repo maturity:** <span style="color: lightblue;">
The firm financed most of its securities position through longer maturity repurchase agreements, to avoid the risk of having to sell securities at a loss to meet margin calls, and maintained their average maturity longer (without letting the average shorten).
</span>

- **Equity redemption:**

<span style="color: lightblue;">The firm had a staggered redemption schedule to keep an average maturity and avoid dates when large amount of equity capital could be withdrawn.

</span>

- **Loan access:**
<span style="color: lightblue;">
The firm secured $230 million in unsecured term loans with three-year maturities and negotiated a $700 million revolving line of credit with a syndicate of 25 banks. To ensure consistent access, LTCM removed discretionary clauses, such as Material Adverse Change (MAC) clauses, from the credit agreement. Instead, a clear covenant restricted access only if the Fund's NAV fell by 50% or more in a year, providing predictable and reliable funding terms
</span>


### 1.4 LTCM is largely in the business of selling liquidity and volatility. Describe how LTCM accounts for liquidity risk in their quantitative measurements.

 
 - <span style="color: lightblue;"> As for its assets and liabilities, "The firm viewed liquidity-related risk as being relatively short-lived (less than a year), and considered it mostly a source of opportunities." Over short horizons, to factor in the trader's liquidity needs, the firm assumed that all portfolio trades were always positively correlated, regardless of which side of the trade the Fund was on, which was a conservative approach.
 
 - <span style="color: lightblue;"> Additionally, the firm financed most of its securities position through longer maturity repo, to avoid the risk of having to sell securities at a loss to meet margin calls.
 
 - <span style="color: lightblue;"> LTCM also considered the use of two-way mark to market for contractuals, which allowed it to  use the collateral inflow related to one position to outflow the outflow needed for another position, which was not common in the industry between a bank and its clients. The company tried to structure its trades so that the only flow of capital would be related to net mark-to-market changes on its positions.

 - <span style="color: lightblue;"> As for its equity liquidity risk (risk that the investors request for withdraw in an inopportune time), the firm had a staggered redemption schedule to keep an average maturity and avoid dates when large amount of equity capital could be withdrawn.
 </span>


### 1.5 Is leverage risk currently a concern for LTCM?

 <span style="color: lightblue;"> No, leverage does not mean much in LTCM business, since it could have a leverage of, say, 10x and a totally hedged offsetting position with zero risk through its long & short trades. 
 </span>


### 1.6 Many strategies of LTCM rely on converging spreads. LTCM feels that these are almost win/win situations because of the fact that if the spread converges, they make money. If it diverges, the trade becomes even more attractive, as convergence is still expected at a future date. What is the risk in these convergence trades?

 <span style="color: lightblue;">
 The risk in these convergence trades are due to timing issues and potential extreme divergence, leading to mark-to-market losses and margin calls, which strain liquidity. During market stress, spreads can widen further, forcing unfavorable liquidations. Liquidity risk increases as counterparties demand higher haircuts, and unexpected correlations or flawed models amplify losses. LTCM's high leverage magnifies these risks, making the strategy vulnerable to systemic shocks, as seen in the 1998 Russian debt crisis, when spreads diverged instead of converging, causing significant financial strain.
</span>

---

# Section 2: LTCM Risk Decomposition

### 2.0 Data Collection
- On Canvas, find the data file, “ltcm exhibits data.xlsx”. Get the gross and net (total) returns of LTCM from “Exhibit 2”.

<span style="color: lightblue;"> </span>


In [1]:
# Import Libraries:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re
import datetime
from typing import Union, List, Callable

import warnings
warnings.filterwarnings("ignore")

pd.options.display.float_format = "{:,.4f}".format
pd.set_option('display.width', 200)
pd.set_option('display.max_columns', 30)

import statsmodels.api as sm
from scipy.stats import t

import os
import sys

parent_path = os.path.dirname(os.getcwd()) # Get parent path (if using .ipynb file)
# parent_path = os.path.dirname(os.path.dirname(os.path.abspath(__file__))) # Get parent path (if using .py file)
os.chdir(parent_path) # Set parent path as working directory (for reading and writing files)
sys.path.insert(0, parent_path) # Add parent path to system path (for importing modules)

import utils.portfolio_management_functions as pm

In [53]:
# Check data in the file (sheets, columns, data):
INFILE_LTCM = "data/ltcm_exhibits_data.xlsx"
try:
    pm.read_excel_default(INFILE_LTCM, print_sheets = True)
except FileNotFoundError as e:
    print(f'{e}.\nCheck file in {parent_path}')

Sheet name: Copyright
Columns: Long-Term Capital - C
                   Long-Term Capital - C
0  Harvard Business School Case #200-009
1                  Case Software #XLS035
2                                    NaN
----------------------------------------------------------------------


Sheet name: Exhibit 2
Columns: Exhibit 2     Fund Capital and Monthly Returns (June, 1994 to July, 1998), Unnamed: 1, Unnamed: 2, Unnamed: 3, Unnamed: 4
  Exhibit 2     Fund Capital and Monthly Returns (June, 1994 to July, 1998)                Unnamed: 1                  Unnamed: 2                Unnamed: 3                Unnamed: 4
0                                                NaN                                              NaN                         NaN                       NaN                       NaN
1                                                NaN                         Fund Capital ($billions)  Gross Monthly Performancea  Net Monthly Performanceb  Index of Net Performance
2          

In [105]:
# Import data from the file:
exhibit_2 = pd.read_excel(INFILE_LTCM, sheet_name = 'Exhibit 2', skiprows = 2, index_col=0)

exhibit_2 = exhibit_2.dropna(axis=0)
exhibit_2.index.name = 'date'


gross_total_return = exhibit_2['Gross Monthly Performancea']
gross_total_return.name = 'gross_total_return'

net_total_return = exhibit_2['Net Monthly Performanceb']
net_total_return.name = 'net_total_return'

gross_total_return.index = pd.to_datetime(gross_total_return.index)
net_total_return.index = pd.to_datetime(net_total_return.index)

print(gross_total_return.head(3))
print()
print(net_total_return.tail(3))

date
1994-03-01   -0.0110
1994-04-01    0.0140
1994-05-01    0.0680
Name: gross_total_return, dtype: float64

date
1998-05-01   -0.0640
1998-06-01   -0.1010
1998-07-01    0.0000
Name: net_total_return, dtype: float64


- Get the returns on SPY as well as the risk-free rate from the file, “gmo analysis data”.
<span style="color: lightblue;"> </span>


In [55]:
# Check data in the file (sheets, columns, data):
INFILE_GMO = "data/gmo_analysis_data.xlsx"
try:
    pm.read_excel_default(INFILE_GMO, print_sheets = True)
except FileNotFoundError as e:
    print(f'{e}.\nCheck file in {parent_path}')

Sheet name: info
Columns: Ticker, Description
        Ticker                               Description
0  SPX DVD YLD          S&P 500 Index for dividend yield
1      SPX P/E  S&P 500 Index for price-earning multiple
2   TNote 10YR                    10-year Treasury yield
----------------------------------------------------------------------


Sheet name: signals
Columns: date, SPX DVD YLD, SPX P/E, TNote 10YR
        date  SPX DVD YLD  SPX P/E  TNote 10YR
0 1996-12-31       1.9651  19.6873      6.4180
1 1997-01-31       1.8455  20.8856      6.4940
2 1997-02-28       1.8502  21.0116      6.5520
----------------------------------------------------------------------


Sheet name: risk-free rate
Columns: date, TBill 3M
        date  TBill 3M
0 1989-12-31    0.0764
1 1990-01-31    0.0774
2 1990-02-28    0.0790
----------------------------------------------------------------------


Sheet name: total returns
Columns: date, SPY, GMWAX, GMGEX
        date     SPY  GMWAX  GMGEX
0 1993-02-28  

In [107]:
# Import data from the file:
spy_total_returns = pd.read_excel(INFILE_GMO, sheet_name='total returns',index_col=0)['SPY']
spy_total_returns.name = 'spy_total_returns'
risk_free_returns = pd.read_excel(INFILE_GMO, sheet_name='risk-free rate',index_col=0)['TBill 3M'] * (1/12)
risk_free_returns.name = 'risk_free_returns'
spy_total_returns.index = pd.to_datetime(spy_total_returns.index).to_period('M').to_timestamp()
risk_free_returns.index = pd.to_datetime(risk_free_returns.index).to_period('M').to_timestamp()

print(spy_total_returns.head(3))
print()
print(risk_free_returns.tail(3))

date
1993-02-01    0.0107
1993-03-01    0.0224
1993-04-01   -0.0256
Freq: MS, Name: spy_total_returns, dtype: float64

date
2024-07-01   0.0042
2024-08-01   0.0039
2024-09-01   0.0038
Freq: MS, Name: risk_free_returns, dtype: float64


### 2.1 Summary stats.
#### (a) For both the gross and net series of LTCM excess returns, report the mean, volatility, and Sharpe ratios. (Annualize them.)

#### (b) Report the skewness, kurtosis, and (historic) VaR(.05).

#### (c) Comment on how these stats compare to SPY and other assets we have seen. How much do they differ between gross and net?
<span style="color: lightblue;"> 
When compared to the SPY total return, the LTCM gross total return has a higher mean, higher sharpe, slightly higher volatility, higher skewness, higher kurtoisis (positive, while SPY is negative), and lower historical VaR.


The net total_return, on the other hand, when compared to the SPY total return, has a lower mean, a very similar volatility, lower Sharpe, lower skewness, higher kurtosis, and lower historical VaR. 

The gross and net returns differ mostly in the mean annualized return (almost 9% lower for the net returns, meaning that the fees and expenses are quite high). While it also has a annualized vol 2% lower, the fees and expenses lead to a lower sharpe ratio). The net returns distribution also have fatter tail and more negative skewness.

In summary, LTCM’s gross returns suggest impressive performance, significantly exceeding SPY's in both absolute and risk-adjusted terms. However, the drop to net returns highlights how costs substantially eroded its profitability, bringing its Sharpe ratio below SPY's. This demonstrates the importance of accounting for fees and other operational expenses when evaluating investment strategies.
</span>

In [ ]:
# Calculate LTCM summary return statistics:
returns_temp = pd.concat([gross_total_return, net_total_return],axis=1)
returns_temp = pd.merge(returns_temp, spy_total_returns.to_frame(), how='left', left_index=True, right_index=True)

ltcm_stats = pm.calc_returns_statistics(returns=returns_temp,
                                       rf_returns = risk_free_returns,
                                       provided_excess_returns=False,
                                       annual_factor=12,
                                       var_quantile=0.05,
                                       tail_risks=True,
                                       keep_columns=['Annualized Mean', 'Annualized Vol', 'Annualized Sharpe', 'Skewness', 'Kurtosis', 'Historical VaR'])

display(ltcm_stats)

Annualized Mean  Annualized Vol  Annualized Sharpe  Skewness  Excess Kurtosis  Historical VaR (5.0%)  Annualized Historical VaR (5.0%)
gross_total_return           0.2939          0.1364             1.7847   -0.2964           1.5694                -0.0264                           -0.0915
net_total_return             0.2072          0.1119             1.3997   -0.8179           2.9055                -0.0224                           -0.0776
spy_total_returns            0.2258          0.1129             1.5526   -0.4372          -0.3609                -0.0431                           -0.1494

### 2.2 Using the series of net LTCM excess returns, denoted $\tilde{r}^{LTCM}$, estimate the following regression:
$$
\tilde{r}_{LTCM,t} = \alpha + \beta_m \tilde{r}_{m,t} + \epsilon_t
$$
#### (a) Report $\alpha$ and $\beta_m$. Report the $R^2$ stat.

#### (b) From this regression, does LTCM appear to be a “closet indexer”?
<span style="color: lightblue;"> 
The market beta is very low (0.14), while the annualized alpha is significant at a 1% level, suggesting that LTCM is not a closet indexer. The R-squared is very low (0.02), which suggests that the market factor does not explain much of the variation in LTCM's returns.
</span>

#### (c) From the regression, does LTCM appear to deliver excess returns beyond the risk premium we expect from market exposure?
<span style="color: lightblue;">
Yes, the market risk premium explain only 2% of the variation in LTCM's returns, while the alpha is significant at a 1% level, suggesting that LTCM delivers excess returns beyond the risk premium we expect from market exposure.
</span>

In [ ]:
# Calculate the regression of the net returns of LTCM on the S&P 500 returns:
ltcm_net_excess_returns = pd.merge(net_total_return, risk_free_returns, how='left', left_index=True, right_index=True)
ltcm_net_excess_returns = ltcm_net_excess_returns.assign(net_excess_return = ltcm_net_excess_returns['net_total_return'] - ltcm_net_excess_returns['risk_free_returns'] * (1/12))
ltcm_net_excess_returns = ltcm_net_excess_returns['net_excess_return']

ltcm_net_excess_regres = pm.calc_regression(
    Y = ltcm_net_excess_returns,
    X = spy_total_returns,
    intercept=True,
    annual_factor=12)

display(ltcm_net_excess_regres)

y has lenght 53 and X has lenght 381. Joining y and X by y.index...


net_excess_return
R-Squared                               0.0196
Observed Mean                           0.0169
Observed Std Dev                        0.0323
RSE                                     0.0323
MAE                                     0.0232
Alpha                                   0.0143
Annualized Alpha                        0.1716
P-Value (Alpha)                         0.0075
Beta (spy_total_returns)                0.1389
P-Value (spy_total_returns)             0.3169
Tracking Error                          0.0320
Annualized Tracking Error               0.1108

### 2.3 Let’s check for non-linear market exposure. Run the following regression on LTCM’s net excess returns:
$$
\tilde{r}_{LTCM,t} = \alpha + \beta_1 \tilde{r}_{m,t} + \beta_2 (\tilde{r}_{m,t})^2 + \epsilon_t
$$
#### (a) Report $\beta_1$, $\beta_2$, and the $R^2$ stat.

#### (b) Does the quadratic market factor do much to increase the overall LTCM variation explained by the market?
<span style="color: lightblue;">
No, the quadratic market factor does not increase the overall LTCM variation explained by the market, as the R-squared is still very low (0.026).
</span>

#### (c) From the regression evidence, does LTCM’s market exposure behave as if it is long market options or short market options?
<span style="color: lightblue;">
From the regression, it appears to be neutral, since the beta on market returns is not statistically significant (p-value of 0.26). However, if it were significant, then the positive beta would suggest that LTCM is long market options.

</span>

#### (d) Should we describe LTCM as being positively or negatively exposed to market volatility?
<span style="color: lightblue;">
We should describe LTCM as being neutral to market volatility, since the beta on the squared market returns is not statistically significant (p-value of 0.58). However, if it were significant, then the negative beta would suggest that LTCM is negatively exposed to market volatility. The large magnitude of the beta has to do with the fact that the returns are small, which becomes even smaller when squared, so the beta must be large to compensate for the squared returns and fit the model.

</span>

In [ ]:
# Non-linear regression (with squared excess returns):
spy_non_linear_excess_returns = spy_total_returns.to_frame().assign(spy_squared_excess_returns = spy_total_returns**2)

ltcm_net_excess_non_lin_regres = pm.calc_regression(
    Y = ltcm_net_excess_returns,
    X = spy_non_linear_excess_returns,
    intercept=True,
    annual_factor=12)

display(ltcm_net_excess_non_lin_regres)

y has lenght 53 and X has lenght 381. Joining y and X by y.index...


net_excess_return
R-Squared                                        0.0258
Observed Mean                                    0.0169
Observed Std Dev                                 0.0323
RSE                                              0.0325
MAE                                              0.0232
Alpha                                            0.0162
Annualized Alpha                                 0.1949
P-Value (Alpha)                                  0.0118
Beta (spy_total_returns)                         0.1880
P-Value (spy_total_returns)                      0.2559
Beta (spy_squared_excess_returns)               -2.0519
P-Value (spy_squared_excess_returns)             0.5764
Tracking Error                                   0.0319
Annualized Tracking Error                        0.1104

### 2.4 Let’s try to pinpoint the nature of LTCM’s nonlinear exposure. Does it come more from exposure to up-markets or down-markets? Run the following regression on LTCM’s net excess returns:

$$
\tilde{r}_{LTCM,t} = \alpha + \beta \tilde{r}_{m,t} + \beta_u \max (\tilde{r}_{m,t} - k_1, 0) + \beta_d \max (k_2 - \tilde{r}_{m,t}, 0) + \epsilon_t
$$
Where $k_1 = 0.03$ and $k_2 = -0.03$ (approximately one standard deviation of $\tilde{r}_m$).

#### (a) Report $\beta$, $\beta_u$, $\beta_d$, and the $R^2$ stat.

#### (b) Is LTCM long or short the call-like factor? And the put-like factor?
<span style="color: lightblue;">
Neither of the betas to the call-like or put-like factors are statistically significant, so we cannot say that LTCM is either long or short the call-like or put-like factor.
However, if the betas were significant, then the negative beta to the call-like factor would suggest that LTCM is short the call-like factor, while the positive beta to the put-like factor would suggest that LTCM is also short the put-like factor.
</span>

#### (c) Which factor moves LTCM more, the call-like factor, or the put-like factor?
<span style="color: lightblue;">
Neither of the estimated betas are statistically significant, so we cannot say which factor moves LTCM more.
However, if they were, for a one standard deviation move in the call-like factor, LTCM would be moved more by the put-like factor, since the absolute value of the beta to the put-like factor is larger than the absolute value of the beta to the call-like factor. That means that large market movements would lead to a significant underperformance of LTCM. Thus, it seems that LTCM is short market volatility, also because those betas are larger in magnitude than the market beta.
</span>

#### (d) In the previous problem, you commented on whether LTCM is positively or negatively exposed to market volatility. Using this current regression, does this volatility exposure come more from being long the market’s upside? Short the market’s downside? Something else?

<span style="color: lightblue;">
As explained in the previous answer, this regression would imply that LTCM is negatively exposed to market volatility, which is in lign with the previous problem answer. This volatility exposure comes more from being short the market’s downside, since the beta to the put-like factor is larger in magnitude than the beta to the call-like factor.
</span>

In [ ]:
# Non-linear regression with threshold (call-like and put-like options):
THRESHOLD = 0.03
spy_non_linear_ud_excess_returns = spy_total_returns.to_frame().assign(spy_call = np.maximum(spy_total_returns-THRESHOLD, 0))
spy_non_linear_ud_excess_returns = spy_non_linear_ud_excess_returns.assign(spy_put = np.maximum(-THRESHOLD-spy_total_returns, 0))

ltcm_net_excess_non_lin_regres = pm.calc_regression(
    Y = ltcm_net_excess_returns,
    X = spy_non_linear_ud_excess_returns,
    intercept=True,
    annual_factor=12)

display(ltcm_net_excess_non_lin_regres)

y has lenght 53 and X has lenght 381. Joining y and X by y.index...


net_excess_return
R-Squared                               0.0552
Observed Mean                           0.0169
Observed Std Dev                        0.0323
RSE                                     0.0323
MAE                                     0.0224
Alpha                                   0.0111
Annualized Alpha                        0.1333
P-Value (Alpha)                         0.0935
Beta (spy_total_returns)                0.4973
P-Value (spy_total_returns)             0.1016
Beta (spy_call)                        -0.7665
P-Value (spy_call)                      0.2101
Beta (spy_put)                          1.5704
P-Value (spy_put)                       0.2904
Tracking Error                          0.0314
Annualized Tracking Error               0.1088

---

# Section 3: The FX Carry Trade

Find an Excel data file, `"data/fx_rf_data.xlsx"`. The file has two sets of data:
- **“risk-free rates”** - daily quotes of annualized risk-free rates across currencies.
- **“exchange rates”** - daily quotes of spot FX rates expressed as direct quotes to the USD.

### Data Processing

- For risk-free rate data, $r_{t,t+1}^{f,i}$, the rate is known and reported in the data at time $t$. Thus, at any given date $t$, the data file is reporting both $S_i^t$ and $r_{t,t+1}^{f,i}$.

- The theory says to use log risk-free rates. You have the risk-free rate in levels: use the following equation to convert them:
$$
r_{t,t+1}^{f,i} = \ln(1 + r_{t,t+1}^{f,i})
$$

- The theory says to use log spot FX prices. You have the FX prices in levels, so directly take their logarithms:
$$
s_i^t = \ln(S_i^t)
$$


In [128]:
# Check data in the file (sheets, columns, data):
INFILE_CARRY = "data/fx_rf_data.xlsx"
try:
    pm.read_excel_default(INFILE_CARRY, print_sheets = True)
except FileNotFoundError as e:
    print(f'{e}.\nCheck file in {parent_path}')

Sheet name: risk-free rates
Columns: date, USD, JPY, EUR, GBP, MXN, CHF
        date    USD    JPY    EUR    GBP    MXN    CHF
0 2008-01-01 0.0440 0.0048 0.0425 0.0570 0.0750 0.0206
1 2008-01-02 0.0416 0.0048 0.0360 0.0547 0.0750 0.0206
2 2008-01-03 0.0435 0.0048 0.0394 0.0560 0.0750 0.0215
----------------------------------------------------------------------


Sheet name: exchange rates
Columns: date, JPY, EUR, GBP, MXN, CHF
        date    JPY    EUR    GBP    MXN    CHF
0 2008-01-01 0.0090 1.4592 1.9864 0.0918 0.8824
1 2008-01-02 0.0091 1.4715 1.9808 0.0916 0.8933
2 2008-01-03 0.0091 1.4750 1.9710 0.0919 0.9002
----------------------------------------------------------------------




In [235]:
# Import data from the file:

# Exchange rates:
exchange_rates = pd.read_excel(INFILE_CARRY, sheet_name='exchange rates',index_col=0) 
risk_free_rates = pd.read_excel(INFILE_CARRY, sheet_name='risk-free rates',index_col=0)

# Transform to log rates and log prices:
log_exchange_rates = np.log(exchange_rates)
log_ALL_risk_free_rates = np.log(1 + risk_free_rates)


log_FX_rate_returns = log_exchange_rates.diff().iloc[1:]

log_USD_risk_free_rate = log_ALL_risk_free_rates.loc[:,'USD'].iloc[1:]
log_risk_free_rates = log_ALL_risk_free_rates[log_exchange_rates.columns].iloc[1:]

display(log_FX_rate_returns.head(3))
display(log_USD_risk_free_rate.head(3))
display(log_risk_free_rates.head(3))


JPY     EUR     GBP     MXN    CHF
date                                            
2008-01-02 0.0179  0.0084 -0.0028 -0.0023 0.0123
2008-01-03 0.0032  0.0024 -0.0050  0.0027 0.0077
2008-01-04 0.0066 -0.0005  0.0015 -0.0036 0.0022

date
2008-01-02   0.0407
2008-01-03   0.0425
2008-01-04   0.0413
Name: USD, dtype: float64

JPY    EUR    GBP    MXN    CHF
date                                         
2008-01-02 0.0048 0.0354 0.0533 0.0723 0.0204
2008-01-03 0.0048 0.0386 0.0545 0.0723 0.0213
2008-01-04 0.0048 0.0398 0.0535 0.0723 0.0207

### 3.1 The Static Carry Trade
#### Define the log return of holding the foreign currency using log values of the risk-free rate and log values of the FX rates:
$$
r_{t+1}^i \equiv s_{t+1}^i - s_t^i + r_{t,t+1}^{f,i}
$$

Then the excess log return relative to USD is expressed as:
$$
\tilde{r}_{t+1}^i \equiv s_{t+1}^i - s_t^i + r_{t,t+1}^{f,i} - r_{t,t+1}^{f, USD}
$$

#### For each foreign currency, $i$, calculate the excess log return series, $\tilde{r}_{t+1}^i$. Report the following stats (based on the excess log returns) and annualize them:
1. Mean
2. Volatility
3. Sharpe Ratio

#### What differences do you see across currencies?
<span style="color: lightblue;">
Using the above definition, JPY, EUR and GPB had similar negative annualized carry mean between -2% and -3%, with 9-10% annualized volatility.
MXN and CHF had positive annualized mean return of 1.13% and 0.27%, respectively, with annualized vol of 13% and 10.7%.
</span>

In [236]:
# Define the carry returns:
log_currency_returns = log_FX_rate_returns + log_risk_free_rates/252 # Daily excess returns
log_excess_currency_returns = log_currency_returns.subtract(log_USD_risk_free_rate/252, axis=0) 

In [238]:
# Calculate the summary statistics of the carry returns:
currency_returns_stats = pm.calc_returns_statistics(returns=log_excess_currency_returns,
                                                    annual_factor=252,
                                                    provided_excess_returns=True,
                                                    keep_columns=['Annualized Mean', 'Annualized Vol', 'Annualized Sharpe'])
print('Excess Currency Log Returns Statistics:')
display(currency_returns_stats)

Excess Currency Log Returns Statistics:


Annualized Mean  Annualized Vol  Annualized Sharpe
JPY          -0.0294          0.0991            -0.2962
EUR          -0.0235          0.0901            -0.2614
GBP          -0.0249          0.0957            -0.2598
MXN           0.0113          0.1300             0.0869
CHF           0.0027          0.1069             0.0248

### 3.2 Implications for UIP
#### (a) Do any of these stats contradict the (log version) of Uncovered Interest Parity (UIP)?
<span style="color: lightblue;">
Yes, all of them contradict, because the UIP states that the expected returns of carrying foreign currency should be equal to the US risk-free rate. However, the data shows that the expected returns are not equal to the US risk-free rate. However, the MXN excess return (1.13%) was close to the US risk-free rate (1.21%).
</span>

#### (b) A long position in which foreign currency offered the best Sharpe ratio over the sample?
<span style="color: lightblue;">
A long position in MXN offered the best Sharpe ratio over the sample, with a Sharpe ratio of 0.09.
</span>

#### (c) Are there any foreign currencies for which a long position earned a negative excess return (in USD) over the sample?  
<span style="color: lightblue;">
FOr JPY, EUR and GBP, a long position earned a negative excess return (in USD) over the sample.
</span>

In [240]:
print(f"{log_USD_risk_free_rate.iloc[1:].mean():.4f}")

0.0121


### 3.3 Predicting FX
For each foreign currency, test whether interest-rate differentials can predict growth in the foreign-exchange rate by estimating the following forecasting regression:
$$
s_{t+1}^i - s_t^i = \alpha^i + \beta^i \left( r_{t,t+1}^{f,USD} - r_{t,t+1}^{f,i} \right) + \epsilon_{t+1}^i
$$
Where $r_{f,i}$ denotes the risk-free rate of currency $i$, and $s_i$ denotes the FX rate for currency $i$. Both $r_{t,t+1}^{f,USD}$ and $s_t$ are determined at time $t$.

#### (a) Create a table with the following:
- **Columns**: Corresponding to a different currency regression.
- **Rows**: Report $\alpha^i$, $\beta^i$, and $R^2$ estimates for each regression.

#### (b) Suppose the foreign risk-free rate increases relative to the US rate:
- For which foreign currencies would we predict a relative strengthening of the USD in the following period? 

<span style="color: lightblue;">
For MXN, because it has positive beta.
</span>

- For which currencies would we predict relative weakening of the USD in the following period?

<span style="color: lightblue;">
For JPY, EUR, and GBP, because they have negative betas.
</span>

- This FX predictability is strongest in the case of which foreign currency?

<span style="color: lightblue;">
For GBP, because it has the higherst R-squared. However, it is still negligible, at 0.13%
</span>


In [241]:
log_risk_free_diff = log_risk_free_rates.subtract(log_USD_risk_free_rate, axis=0)
# now individually:

consolidated_predict_regress = pd.DataFrame(columns=log_FX_rate_returns.columns,
                                            index=['Annualized Alpha', 'Beta', 'R-Squared'])
for currency in log_FX_rate_returns.columns:
    predict_fx_regression = pm.calc_regression(Y = log_FX_rate_returns[currency],
                                                X = log_risk_free_diff[currency],
                                                intercept=True,
                                                annual_factor=252)
    
    consolidated_predict_regress.loc['Annualized Alpha', currency] = predict_fx_regression.loc['Annualized Alpha'].values[0]
    consolidated_predict_regress.loc['Beta', currency] = predict_fx_regression.loc[predict_fx_regression.index.str.contains(r'Beta \(')].values[0]
    consolidated_predict_regress.loc['R-Squared', currency] = predict_fx_regression.loc['R-Squared'].values[0]

display(consolidated_predict_regress.style.format("{:.4f}"))

### 3.4 The Dynamic Carry Trade
Use the forecasting regression results to write $\mathbb{E}\left[ \tilde{r}_i^{t+1} \right]$ as a function of the interest-rate differential as well as $\alpha$ and $\beta$:
$$
\mathbb{E}[s_{t+1} - s_t] = \alpha^i + \beta^i \left( r_{t,t+1}^{f,USD} - r_{t,t+1}^{f,i} \right)
$$

From the definition of excess (log) returns on FX:
$$
\tilde{r}_{t+1}^i = s_{t+1} - s_t - \left( r_{t,t+1}^{f,USD} - r_{t,t+1}^{f,i} \right)
$$
Rearranging implies the following forecast for excess log returns:
$$
\mathbb{E}[\tilde{r}_i^{t+1}] = \alpha^i + (\beta^i - 1) \left( r_{t,t+1}^{f,USD} - r_{t,t+1}^{f,i} \right)
$$


Use your regression estimates from Problem 3 along with the formula above to:
#### (a) Calculate:
   - The fraction of months for which the estimated FX risk premium is positive. That is, for each currency $i$, calculate how often this condition holds in the time-series:
   $$\mathbb{E}[\tilde{r}_{t+1}^i] > 0$$
   
   <span style="color: lightblue;"> </span>

#### (b) Identify:
   - Which currencies most consistently have a positive FX risk premium? 
   
   <span style="color: lightblue;">
   CHF has the highest fraction of days for which the estimated FX risk premium is positive, at 0.9934.
   </span>

   - For which currencies does the FX risk premium most often go negative?
   
   <span style="color: lightblue;">
   For JPY, most of days it has a negative risk premium (only 19.5% of days it is positive). For EUR, GBP and MXN, the risk premium was always negative.
   </span>

#### (c) Explain:
   How could these conditional risk premia be used to improve the static carry trade returns calculated in Problem 3.2? <span style="color: lightblue;">
   No, because they have little explanation power.</span>

In [250]:
# Get fraction of months for which the estimated FX risk premium is positive:

fx_premium_positive = pd.DataFrame(index=['Days (positive premium)', 'Days (total)', 'Frequency (positive premium)'], columns=log_FX_rate_returns.columns)
for currency in log_FX_rate_returns.columns:
    expected_excess_return = (consolidated_predict_regress.loc['Annualized Alpha', currency] +
                              (consolidated_predict_regress.loc['Beta', currency] - 1)* log_risk_free_diff[currency])
    
    fx_premium_positive.loc['Days (positive premium)', currency] = int((expected_excess_return > 0).sum())
    fx_premium_positive.loc['Days (total)', currency] = int(len(expected_excess_return))
    fx_premium_positive.loc['Frequency (positive premium)', currency] = "{:.2%}".format((expected_excess_return > 0).mean())

fx_premium_positive.style

    